library/psa_crypto.c

```diff
@@ -7108,6 +7115,13 @@ psa_status_t psa_crypto_init(void)
         return PSA_SUCCESS;
     }
 
+    /* Init drivers */
+    status = psa_driver_wrapper_init();
```

Comment:
Currently initializing just the drivers is not supported (`psa_driver_wrapper_init()` is a private function), the only thing that's supported is initializing everything with `psa_crypto_init()` which already has a guard, so I think that's OK.

(Note: https://github.com/Mbed-TLS/mbedtls/pull/6636 will change that but it's not in yet.)

Reply:
What's the gain of introducing `drivers_initialized` then? I assumed that it should support partial `psa_crypto_init` success:
- If `psa_crypto_init` succeeds, then indeed, `global_data.initialized` is equal to 1 and there's no need to check `drivers_initialized`;
- If `psa_crypto_init` fails after `psa_driver_wrapper_init` was called succesfully then the state is inconsistent and we could prevent a second call to drivers that were already initialized, during a second call to `psa_crypto_init`;
If that's not the point and we don't want to support partial initialization, then why add `drivers_initialized` at all?

Reply:
Ah, excellent question! Indeed in the context of this PR there's not point yet, I'm just anticipating on #7155 where the entropy module will start using MD, with the goal of having it work when hashes are provided by drivers. The entropy module is first used at the "init RNG" stage in `psa_crypto_init()`, so when it will call MD, `global_data.initialized` will still be 0 (it's only changed to 1 at the end of the function), but we'll want MD to dispatch to drivers nonetheless (which is also the reason for the change of ordering).

Now that you mention it, my PR structure is not clean: in this PR indeed I shouldn't have changed the order or introduced `drivers_initialized`, but instead use `global_data.initialized` in `psa_can_do_hash()`, and left the rest for the PR about the entropy module. I'm afraid I got a bit ahead of myself here.

Reply:
Ah, now it makes sense! Thanks for explaining. I still think that we could optimize to not call init on drivers if they are already initialized, but it's a minor.